src/plugins/mvcc_delete_plugin.cpp

```diff
@@ -0,0 +1,184 @@
+
+      if (lowest_snapshot_commit_id.has_value()) {
+        conflicting_transactions = chunk->get_cleanup_commit_id().value() > lowest_snapshot_commit_id.value();
+      }
+
+      if (!conflicting_transactions) {
+        _delete_chunk_physically(table, table_and_chunk_id.second);
+        _physical_delete_queue.pop();
+      }
+    }
+  }
+}
+
+bool MvccDeletePlugin::_try_logical_delete(const std::string& table_name, const ChunkID chunk_id) {
+  const auto& table = StorageManager::get().get_table(table_name);
+  const auto& chunk = table->get_chunk(chunk_id);
+
+  Assert(chunk != nullptr, "Chunk does not exist. Logical Delete can not be applied.");
+  Assert(chunk_id < (table->chunk_count() - 1),
+         "MVCC Logical Delete should not be applied on the last/current mutable chunk.");
+
+  // Create temporary referencing table that contains the given chunk only
+  auto transaction_context = TransactionManager::get().new_transaction_context();
+  auto gt = std::make_shared<GetTable>(table_name);
+  gt->set_transaction_context(transaction_context);
+
+  // Include all ChunksIDs of current table except chunk_id for pruning in GetTable
+  std::vector<ChunkID> excluded_chunk_ids(table->chunk_count() - 1);
+  std::iota(excluded_chunk_ids.begin(), excluded_chunk_ids.begin() + chunk_id, 0);
+  std::iota(excluded_chunk_ids.begin() + chunk_id, excluded_chunk_ids.end(), chunk_id + 1);
+
+  gt->set_excluded_chunk_ids(excluded_chunk_ids);
+  gt->execute();
+
+  // Validate temporary table
+  auto validate_table = std::make_shared<Validate>(gt);
+  validate_table->set_transaction_context(transaction_context);
+  validate_table->execute();
+
+  // Use Update operator to delete and re-insert valid records in chunk
+  // Pass validate_table into Update operator twice since data will not be changed.
+  auto update_table = std::make_shared<Update>(table_name, validate_table, validate_table);
+  update_table->set_transaction_context(transaction_context);
+  update_table->execute();
+
+  // Check for success
+  if (update_table->execute_failed()) {
+    transaction_context->rollback();
+    return false;
+  } else {
+    // TODO(all): Check for success of commit, currently (2019-01-11) not yet possible (#1393).
+    transaction_context->commit();
+
+    // Mark chunk as logically deleted
+    chunk->set_cleanup_commit_id(transaction_context->commit_id());
+    return true;
+  }
+}
+
+void MvccDeletePlugin::_delete_chunk_physically(const std::shared_ptr<Table>& table, const ChunkID chunk_id) {
+  const auto& chunk = table->get_chunk(chunk_id);
+
+  /**
+   * The chunk should be referenced only by the plugin in methods _delete_chunk_physically()
+   * & _physical_delete_loop() and the table's chunk vector.
+   */
+  Assert(chunk.use_count() == 3,
+         "At this point, the chunk should be referenced only by the plugin in methods _delete_chunk_physically() "
+         "& _physical_delete_loop() and the table's chunk vector.");
```

Comment:
What if a different thread just now looked at the chunk, made its copy, and has not yet decided that it doesn't have to process it as the cleanup commit id is older than its snapshot?

Reply:
Yeah, that's a problem. So, we really need a mutex? It can also fail if get_chunk already returns a nullptr when the chunk is logically deleted.

Reply:
I think we are talking about different things. I am just saying that the use_count might be higher than 3. Is that really a problem though?

Reply:
Such a thread should not take ownership of the chunk. Saving a `chunk_id` is the better approach.